mantis/simulation/routers/solver.py

```diff
@@ -0,0 +1,205 @@
+
+        A = []
+        for exchange in exchanges:
+            n_i = len(exchange.assets)  # Number of tokens in pool (default to 2)
+            A_i = np.zeros((num_assets, n_i))
+            for i, token in enumerate(exchange.assets):
+                A_i[asset_index(token.id), i] = 1
+            A.append(A_i)
+
+        # tendered (given) amount
+        deltas = [
+            cp.Variable(len(exchange.assets), nonneg=True) for exchange in exchanges
+        ]
+
+        # received (wanted) amounts
+        lambdas = [
+            cp.Variable(len(exchange.assets), nonneg=True) for exchange in exchanges
+        ]
+        
+        # Binary value, indicates tx or not for given pool
+        eta = cp.Variable(
+            num_exchanges, boolean=is_micp, nonneg=not is_micp
+        )
+
+        # network trade vector - net amount received over all trades(transfers/exchanges)
+        psi = cp.sum(
+            [A_i @ (LAMBDA - DELTA) for A_i, DELTA, LAMBDA in zip(A, deltas, lambdas)]
+        )
+
+        # Objective is to trade number_of_init_tokens of asset origin_token for a maximum amount of asset objective_token
+        obj = cp.Maximize(
+            psi[asset_index(input.out_token_id)]
+            - cp.sum(eta @ [exchange.fix_cost for exchange in exchanges])
+        )
+
+        # Trading function constraints
+        constrains = [
+            psi + initial_assets >= 0,
+        ]
+
+        for i, exchange in enumerate(exchanges):
+            new_reserve = self.get_new_reserve(deltas[i], lambdas[i], exchange)
+            constrains.extend(exchange.constraints(new_reserve))
+            if not is_micp:
+                constrains.append(eta >= 0)
+                constrains.append(eta <= 15)
+                constrains.append(
+                    cp.sum(deltas[i] @ [asset.price for asset in exchange.assets])
+                    <= eta[i]
+                    * 20
+                    * input.in_amount
+                    * assets[asset_index(input.in_token_id)].price
+                )
+            else:
+                constrains.append(deltas[i] <= eta[i] * exchange.reserves * 0.5)
+
+        # Set up and solve problem
+        prob = cp.Problem(obj, constrains)
+        prob.solve(solver=cp.SCIP, verbose=True)
+
+        print(
+            f"\033[1;91m[{num_exchanges}]Total amount out: {psi.value[asset_index(input.out_token_id)]}\033[0m"
+        )
+
+        for i in range(num_exchanges):
+            print(
+                f"Market {exchanges[i].assets[0]}<->{exchanges[i].assets[1]}, delta: {deltas[i].value}, lambda: {lambdas[i].value}, eta: {eta[i].value}",
+            )
+        if not is_micp:
+            exchanges = [
+                pair[0]
+                for pair in sorted(
+                    zip(exchanges, eta), key=lambda ele: ele[1].value, reverse=True
+                )
+            ][:self.MICP_THRESHOLD]
+            self._solve(exchanges, assets, input)
```

Comment:
interesting, not clear so why MAX eta picked for list of treshhold. I see it is weird value so. Does it work on original simulator?